dbcon/mysql/ha_mcs_datatype.h

```diff
@@ -539,264 +553,28 @@ class WriteBatchFieldMariaDB: public WriteBatchField
-          totalBytes++;
-
-        bytesBefore = totalBytes - 4;
-        break;
-      }
-
-      case 8:
-      {
-        if ((m_type.precision == 18) || (m_type.precision == 15)
-            || (m_type.precision == 13) || (m_type.precision == 11)
-            || (m_type.precision == 9))
-          totalBytes++;
-
-        bytesBefore = totalBytes - 4;;
-        break;
-      }
-
-      case 9:
-      {
-        bytesBefore = totalBytes - 4;;
-        break;
-      }
-
-      case 10:
-      {
-        if ((m_type.precision != 16) && (m_type.precision != 14)
-            && (m_type.precision != 12) && (m_type.precision != 10))
-          totalBytes++;
-
-        bytesBefore = totalBytes - 5;;
-        break;
-      }
-
-      case 11:
-      {
-        if (m_type.precision == 18)
-          totalBytes++;
-
-        bytesBefore = totalBytes - 5;
-        break;
-      }
-
-      case 12:
-      {
-        if ((m_type.precision != 16) && (m_type.precision != 14)
-            && (m_type.precision != 12))
-          totalBytes++;
-
-        bytesBefore = totalBytes - 6;
-        break;
-      }
-
-      case 13:
-      {
-        if (m_type.precision == 18)
-          totalBytes++;
-
-        bytesBefore = totalBytes - 6;
-        break;
-      }
-
-      case 14:
-      {
-        if ((m_type.precision != 16) && (m_type.precision != 14))
-          totalBytes++;
-
-        bytesBefore = totalBytes - 7;
-        break;
-      }
-
-      case 15:
-      {
-        if (m_type.precision == 18)
-          totalBytes++;
-
-        bytesBefore = totalBytes - 7;
-        break;
-      }
-
-      case 16:
-      {
-        if (m_type.precision != 16)
-          totalBytes++;
-
-        bytesBefore = totalBytes - 8;
-        break;
-      }
-
-      case 17:
-      {
-        if (m_type.precision == 18)
-          totalBytes++;
-
-        bytesBefore = totalBytes - 8;
-        break;
-      }
-
-      case 18:
-      {
-        bytesBefore = totalBytes - 8;
-        break;
-      }
-
-      default:
-        break;
-    }
+    uint bytesBefore = numDecimalBytes(m_type.precision - m_type.scale);
+    uint totalBytes = bytesBefore + numDecimalBytes(m_type.scale);
 
     if (nullVal && (m_type.constraintType != CalpontSystemCatalog::NOTNULL_CONSTRAINT))
     {
       fprintf(ci.filePtr(), "%c", ci.delimiter());
       //printf("|");
     }
+    else if (m_type.precision > datatypes::INT64MAXPRECISION)
+    {
+      // TODO MCOL-641 The below else block for narrow decimal
+      // i.e. (m_type.precision <= datatypes::INT64MAXPRECISION)
+      // converts the decimal binary representation in buf directly
+      // to a string, while here, the my_decimal ctor first calls
+      // bin2decimal() on buf, and then we construct the string from
+      // the my_decimal. This approach might be a bit slower than the
+      // narrow decimal approach.
+      my_decimal dec(buf, m_type.precision, m_type.scale);
```

Comment:
The approach looks logical IMHO but I'll ask bar if he can suggest/contribute to this part.